Source/WebCore/html/ValidatedFormListedElement.cpp

```diff
@@ -0,0 +1,495 @@
+        return true;
+    // An event handler can deref this object.
+    HTMLElement& element = asHTMLElement();
+    Ref protectedThis { element };
+    Ref originalDocument { element.document() };
+    auto event = Event::create(eventNames().invalidEvent, Event::CanBubble::No, Event::IsCancelable::Yes);
+    element.dispatchEvent(event);
+    if (!event->defaultPrevented() && unhandledInvalidControls && element.isConnected() && originalDocument.ptr() == &element.document())
+        unhandledInvalidControls->append(this);
+    return false;
+}
+
+bool ValidatedFormListedElement::reportValidity()
+{
+    Vector<RefPtr<ValidatedFormListedElement>> elements;
+    if (checkValidity(&elements))
+        return true;
+
+    if (elements.isEmpty())
+        return false;
+
+    // Needs to update layout now because we'd like to call isFocusable(),
+    // which has !renderer()->needsLayout() assertion.
+    asHTMLElement().document().updateLayoutIgnorePendingStylesheets();
+    if (auto anchor = focusableAnchorElement())
+        focusAndShowValidationMessage(anchor.releaseNonNull());
+    else
+        reportNonFocusableControlError();
+
+    return false;
+}
+
+RefPtr<HTMLElement> ValidatedFormListedElement::focusableAnchorElement()
+{
+    RefPtr<HTMLElement> anchor = anchorElement();
+    if (anchor && anchor->isConnected() && anchor->isFocusable())
+        return anchor;
+    return nullptr;
+}
+
+void ValidatedFormListedElement::focusAndShowValidationMessage(Ref<HTMLElement> anchor)
+{
+    Ref protectedThis { *this };
+    SetForScope isFocusingWithValidationMessageScope(m_isFocusingWithValidationMessage, true);
+
+    // Calling focus() will scroll the element into view.
+    anchor->focus();
+
+    // focus() will scroll the element into view and this scroll may happen asynchronously.
+    // Because scrolling the view hides the validation message, we need to show the validation
+    // message asynchronously as well.
+    callOnMainThread([this, protectedThis, anchor] {
+        updateVisibleValidationMessage(anchor);
+    });
+}
+
+void ValidatedFormListedElement::reportNonFocusableControlError()
+{
+    auto& document = asHTMLElement().document();
+    if (document.frame()) {
+        auto message = makeString("An invalid form control with name='", name(), "' is not focusable.");
+        document.addConsoleMessage(MessageSource::Rendering, MessageLevel::Error, message);
+    }
+}
+
+bool ValidatedFormListedElement::isShowingValidationMessage() const
+{
+    return m_validationMessage && m_validationMessage->isVisible();
+}
+
+bool ValidatedFormListedElement::validationMessageShadowTreeContains(const Node& node) const
+{
+    return m_validationMessage && m_validationMessage->shadowTreeContains(node);
+}
+
+bool ValidatedFormListedElement::isFocusingWithValidationMessage() const
+{
+    return m_isFocusingWithValidationMessage;
+}
+
+void ValidatedFormListedElement::setAncestorDisabled(bool isDisabled)
```

Comment:
I'd call this setDisabledByAncestorFieldset to match the member variable name for clarity.

Reply:
Renamed, nice.